test/e2e/storage/testsuites/readwriteoncepod.go

```diff
@@ -0,0 +1,220 @@
+func InitReadWriteOncePodTestSuite() storageframework.TestSuite {
+	// Only covers one test pattern since ReadWriteOncePod enforcement is
+	// handled through Kubernetes and does not differ across volume types.
+	patterns := []storageframework.TestPattern{storageframework.DefaultFsDynamicPV}
+	return InitCustomReadWriteOncePodTestSuite(patterns)
+}
+
+func (t *readWriteOncePodTestSuite) GetTestSuiteInfo() storageframework.TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *readWriteOncePodTestSuite) SkipUnsupportedTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
+	driverInfo := driver.GetDriverInfo()
+	if !driverInfo.Capabilities[storageframework.CapReadWriteOncePod] {
+		e2eskipper.Skipf("Driver %q doesn't support ReadWriteOncePod - skipping", driverInfo.Name)
+	}
+}
+
+func (t *readWriteOncePodTestSuite) DefineTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
+	var (
+		driverInfo = driver.GetDriverInfo()
+		l          readWriteOncePodTest
+	)
+
+	// Beware that it also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewFrameworkWithCustomTimeouts("read-write-once-pod", storageframework.GetDriverTimeouts(driver))
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	init := func() {
+		l = readWriteOncePodTest{}
+		l.config, l.driverCleanup = driver.PrepareTest(f)
+		l.cs = f.ClientSet
+		l.pods = []*v1.Pod{}
+		l.migrationCheck = newMigrationOpCheck(f.ClientSet, f.ClientConfig(), driverInfo.InTreePluginName)
+	}
+
+	cleanup := func() {
+		var errs []error
+		for _, pod := range l.pods {
+			framework.Logf("Deleting pod %v", pod.Name)
+			err := e2epod.DeletePodWithWait(l.cs, pod)
+			errs = append(errs, err)
+		}
+
+		framework.Logf("Deleting volume %s", l.volume.Pvc.GetName())
+		err := l.volume.CleanupResource()
+		errs = append(errs, err)
+
+		errs = append(errs, storageutils.TryFunc(l.driverCleanup))
+		framework.ExpectNoError(errors.NewAggregate(errs), "while cleaning up resource")
+		l.migrationCheck.validateMigrationVolumeOpCounts()
+	}
+
+	ginkgo.BeforeEach(func() {
+		init()
+		ginkgo.DeferCleanup(cleanup)
+	})
+
+	ginkgo.It("should block a second pod from using an in-use ReadWriteOncePod volume", func() {
+		// Create the ReadWriteOncePod PVC.
+		accessModes := []v1.PersistentVolumeAccessMode{v1.ReadWriteOncePod}
+		l.volume = storageframework.CreateVolumeResourceWithAccessModes(driver, l.config, pattern, t.GetTestSuiteInfo().SupportedSizeRange, accessModes)
+
+		podConfig := e2epod.Config{
+			NS:           f.Namespace.Name,
+			PVCs:         []*v1.PersistentVolumeClaim{l.volume.Pvc},
+			SeLinuxLabel: e2epv.SELinuxLabel,
+		}
+
+		// Create the first pod, which will take ownership of the ReadWriteOncePod PVC.
+		pod1, err := e2epod.MakeSecPod(&podConfig)
+		framework.ExpectNoError(err, "failed to create pod spec")
```

Comment:
nit: it would be nice to include "pod1" / "pod2" in all `ExpectNoError` explanation strings here and below.